plugins/chat/app/services/chat/trash_messages.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible for trashing multiple chat messages
+  # for a channel and ensuring that their client and read state
+  # is updated.
+  #
+  # @example
+  #  Chat::TrashMessages.call(message_ids: [2, 3], channel_id: 1, guardian: guardian)
+  #
+  class TrashMessages
+    include Service::Base
+
+    # @!method call(message_ids:, channel_id:, guardian:)
+    #   @param [Array<Integer>] message_ids
+    #   @param [Integer] channel_id
+    #   @param [Guardian] guardian
+    #   @return [Service::Base::Context]
+
+    contract
+    model :messages
+    policy :invalid_access
+    transaction do
+      step :trash_messages
+      step :destroy_notifications
+      step :update_last_message_ids
+      step :update_tracking_states
+      step :update_thread_reply_cache
+    end
+    step :publish_events
+
+    # @!visibility private
+    class Contract
+      attribute :channel_id, :integer
+      attribute :message_ids, :array
+      validates :channel_id, presence: true
+      validates :message_ids, length: { minimum: 1, maximum: 50 }
+    end
+
+    private
+
+    def fetch_messages(contract:)
+      Chat::Message.includes(chat_channel: :chatable).where(
+        id: contract.message_ids,
+        chat_channel_id: contract.channel_id,
+      )
+    end
+
+    def invalid_access(guardian:, messages:)
+      messages.all? { |message| guardian.can_delete_chat?(message, message.chat_channel.chatable) }
+    end
+
+    def trash_messages(guardian:, messages:)
+      messages.each { |message| message.trash!(guardian.user) }
```

Comment:
This might be a imagineering problem since we're limiting to at most 50 messages to be deleted, but is there a way to do that in one SQL query rather than N? 🤔

Reply:
Ignore that. I think we can re-investigate performance issues if/when they happen. We're pretty safe with the 50 limit.